### PR TITLE
feat(congestion): write stats to a csv file

### PR DESCRIPTION
**Pull request by wacban**
This feature allows dumping the stats to a csv file. In a follow up PR I'll figure out how to plot this data conveniently. I implemented the framework and an example in the SimpleBackpressure strategy. 

By default the model will add a single column `round` and populate it for each round. 
Each strategy needs to populate the head in the init method and populate the values in the compute_chunk method. 

In the simple backpressure strategy the first few rows look like follows:
```
round,shard_0_incoming_queue,shard_0_outgoing_queue,shard_1_incoming_queue,shard_1_outgoing_queue,shard_2_incoming_queue,shard_2_outgoing_queue,shard_3_incoming_queue,shard_3_outgoing_queue
1,0,0,0,0,0,0,0,0
2,20,0,0,0,0,0,0,0
3,40,0,0,0,0,0,0,0
4,57,0,3,0,0,0,0,0
5,73,0,2,0,2,0,0,0
6,89,0,0,0,3,0,1,0
7,105,0,0,0,0,0,4,0
```

### CODE REVIEW

File: tools/congestion-model/src/main.rs [hunk 1]
```diff
@@ -80,19 +83,19 @@ fn strategy(strategy_name: &str, num_shards: usize) -> Vec<Box<dyn CongestionStr
     result
 }
 
-fn parse_workflow_names(workflow_name: &str) -> Vec<String> {
+fn parse_workload_names(workload_name: &str) -> Vec<String> {
```
Comment (wacban) [inline tools/congestion-model/src/main.rs:86]: unrelated renaming, in the previous PR I used workflow instead of workload 🙄

File: tools/congestion-model/src/strategy/mod.rs [hunk 2]
```diff
@@ -9,17 +9,29 @@ mod global_tx_stop;
 mod no_queues;
 mod simple_backpressure;
 
+// The stats writer can be used to dump stats into a CSV file.
+// In the strategy implementation you should write the header in the init method
+// and the individual values in the compute_chunk method. Please use the
+// write_field. The model will take care of writing the record terminator.
+pub type StatsWriter = Option<Box<csv::Writer<std::fs::File>>>;
+
```
Comment (jakmeier) [inline tools/congestion-model/src/strategy/mod.rs:17]: Is it necessary to do leak this evaluation code into strategies? It completely breaks the original architecture, where strategies only care about handling congestion and nothing else.

I would prefer to have all evaluation code in the `.src/evaluation` folder and do it independent of specific strategies. The model is aware of all queues the strategy uses, so it should have all the required information.
If we do it this way, the output will also be standardized across all strategies and hence it's much easier to compare them.

**Comment by wacban** [comment]
Balanced Simple Backpressure

Screenshot 2024-03-07 at 15 38 11

AllToOne SimpleBackpressure

Screenshot 2024-03-07 at 15 37 59

**Comment by wacban** [comment]
A few more stats need to be added as well as stats for other strategies but I'll leave that for later.

File: tools/congestion-model/src/model/mod.rs [hunk 3]
```diff
@@ -63,13 +84,22 @@ impl Model {
...
     pub fn step(&mut self) {
         self.round += 1;
 
+        if let Some(stats_writer) = &mut self.stats_writer {
+            let time = self.start_time + Duration::from_secs(self.round);
+
+            stats_writer.write_field(format!("{:?}", time)).unwrap();
+            stats_writer.write_field(format!("{}", self.round)).unwrap();
+        }
+
```
Comment (jakmeier) [inline tools/congestion-model/src/model/mod.rs:102]: Ideally the `step` function does not need to know about the stats writer at all.
The code running the model (in main.rs) can call a evaluation function after every call to `step()`.
Then we can add all kind of fancy output formatting options in `main.rs` that don't affect how the model runs.
At least that's what I had in mind for the code base architecture.

File: tools/congestion-model/src/model/mod.rs [hunk 4]
```diff
@@ -35,26 +38,44 @@ pub struct Model {
...
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct ShardId(usize);
+pub struct ShardId(pub usize);
 
 impl Model {
     pub fn new(
         mut shards: Vec<Box<dyn CongestionStrategy>>,
         mut producer: Box<dyn Producer>,
+        mut stats_writer: StatsWriter,
```
Comment (jakmeier) [inline tools/congestion-model/src/model/mod.rs:53]: The model constructor should not need an input for the writer.
It should not care about anything other than the strategy and the workload as inputs.

File: tools/congestion-model/src/model/mod.rs [hunk 5]
```diff
@@ -35,26 +38,44 @@ pub struct Model {
...
+        let start_time = Utc::now() - Duration::from_secs(1 * 60 * 60);
         let num_shards = shards.len();
         let shard_ids: Vec<_> = (0..num_shards).map(ShardId).collect();
         let mut queues = QueueBundle::new(&shard_ids);
 
+        if let Some(stats_writer) = &mut stats_writer {
+            stats_writer.write_field("time").unwrap();
+            stats_writer.write_field("round").unwrap();
+        }
+
```
Comment (jakmeier) [inline tools/congestion-model/src/model/mod.rs:68]: this code should be somewhere in `.src/evaluation` and be called by the model runner in main.rs

**Review by jakmeier — CHANGES_REQUESTED** [review]
I love this conceptually! Getting all the data out in a CSV is exactly what we need to solve the evaluation part cleanly. The summary table was a very crude first example of how we could get an output, this is miles better.

However, you are breaking several of the code architecture invariants. :sweat_smile: Not your fault, of course, you couldn't know it. But the way I had it in my mind, I wanted to keep the code base modular with minimal cross-over between the 4 main components (strategy, workload, model, evaluation). You can kind of see it how your changes forced you to pass around the stats writer to many files that shouldn't be affected by such a change.

I've tried to explain what my architectural idea is in in-line comments. Please take a look and let me know if you agree / want to follow my proposed architecture. Or if you think it's too much work or just not a good idea, we can can also work out a different architecture together, I'm open to your suggestions.

File: tools/congestion-model/src/evaluation/mod.rs [hunk 6]
```diff
@@ -72,4 +76,53 @@ impl Model {
...
+    ) {
+        let Some(stats_writer) = stats_writer else { return };
+
+        let time = start_time + Duration::seconds(round as i64);
+
+        stats_writer.write_field(format!("{:?}", time)).unwrap();
+        stats_writer.write_field(format!("{}", round)).unwrap();
+
+        // This is slow and takes up to 10s for the slowest workloads and strategies.
+        let progress = self.progress();
```
Comment (wacban) [inline tools/congestion-model/src/evaluation/mod.rs:114]: @jakmeier Let me know if this is something to be addressed. I think it's fine because it's still ok to run a single strategy and workload and you'd typically want the stats only at the end of dev cycle. One way to improve this would be to add sampling.

File: tools/congestion-model/src/model/queue.rs [hunk 7]
```diff
@@ -4,12 +4,13 @@ use std::collections::VecDeque;
...
 pub struct Queue {
     shard: ShardId,
+    name: String,
     messages: VecDeque<Receipt>,
 }
 
 impl Queue {
-    pub fn new(shard: ShardId) -> Self {
-        Self { shard, messages: VecDeque::new() }
+    pub fn new(shard: ShardId, name: &str) -> Self {
```
Comment (wacban) [inline tools/congestion-model/src/model/queue.rs:12]: This is needed because the queue ids are not particularly human friendly. It however puts extra work for the strategy developers because they'd need to name their queues. It also means that unless the developers come up with consistent naming the queus stats and dashboard panels won't be reusable. I'm open for better suggestions or ideas.

File: tools/congestion-model/src/model/queue_bundle.rs [hunk 8]
```diff
@@ -23,17 +23,17 @@ impl QueueBundle {
         };
 
         for &shard in shards {
-            let mailbox = this.new_queue(shard);
+            let mailbox = this.new_queue(shard, "mailbox");
```
Comment (wacban) [inline tools/congestion-model/src/model/queue_bundle.rs:26]: What the heck is mailbox? :)

**Comment by wacban** [comment]
new impl with new stats

simple backpressure, balanced - unbounded - waiting transactions is ever growing
Screenshot 2024-03-10 at 18 30 52

simple backpressure, all to one - deadlocking - finished transactions goes flat about halfway the evaluation
Screenshot 2024-03-10 at 18 31 43

File: tools/congestion-model/src/model/queue_bundle.rs (hunk 8 above)
Comment (jakmeier) [reply tools/congestion-model/src/model/queue_bundle.rs:26]: Borrowed vocabulary from the actor model

https://www.brianstorti.com/the-actor-model/

The model basically treats each shard as an actor and the implicit incoming receipts queue is the actor's mailbox. And that's no coincidence. I always like to map problems to solutions that already exist.

Not only is the model implemented this way, I think even the real implementation can be thought of this way. Taking inspiration from Erlang, Akka, Actix, and other established actor frameworks could be useful. But it's also quite limited since they generally react to full queues by dropping messages / returning failures (which I still think we can avoid) and they don't operate on a global round-based clock that the blockchain has thanks to block heights. I think those are the main differences in constraints.

Anyway, sorry for the off-topic rambling... "mailbox" as the queue name makes sense in my mind but open to other names :)

File: tools/congestion-model/src/model/mod.rs [hunk 9]
```diff
@@ -38,7 +38,7 @@ pub struct Model {
 }
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub struct ShardId(usize);
+pub struct ShardId(pub usize);
```
Comment (jakmeier) [inline tools/congestion-model/src/model/mod.rs:41]: Is this still needed, or just an artefact from previous code iterations? If it's not necessary, I'd prefer this to stay private.

Making it `pub` is okay for our little model if it makes things easier. But generally speaking, it breaks the intended type safety. While the value is private, we can guarantee that `ShardId` can only be created inside this file. So it's easy to check invariants on it. That's about halve the motivation for wrapping the `usize` in a new type. It's the same pattern I used for `QueueId`, `TransactionId`, and `ReceiptId`.

If you just need read-only access to the number, adding a simple getter method, or maybe `impl From<ShardId> for usize {}` or even `impl Deref for ShardId { type Target = usize }` could do the trick without breaking type properties but still allows reading the number with `id.into()` or `*id`, which seems just as convenient as `id.0`.

**Review by jakmeier — APPROVED** [review]
Thanks @wacban for the changes! I personally like this much better, as it now fits in perfectly with the intended architecture.

This is going to be suuper helpful to analyse and compare different strategies!
